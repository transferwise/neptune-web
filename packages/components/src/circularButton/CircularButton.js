import React, { cloneElement } from 'react';
import Types from 'prop-types';
import classNames from 'classnames';

import { ControlType, Priority } from '../common';
import { typeClassMap, priorityClassMap } from '../button/classMap';

import './CircularButton.css';

const CircularButton = ({ className, children, disabled, icon, priority, type, ...rest }) => {
  const classes = classNames('btn np-btn', typeClassMap[type], priorityClassMap[priority]);

  const iconEl = icon.props.size !== 24 ? cloneElement(icon, { size: 24 }) : icon;

  return (
    <label
      className={`np-circular-btn ${priority} ${type} ${
        (disabled && 'disabled') || ''
      } ${className}`}
    >
      <input
        type="button"
        aria-label={children}
        className={classes}
        disabled={disabled}
        {...rest}
      />
      {iconEl}
      <span className="np-circular-btn__label">{children}</span>
    </label>
  );
};

CircularButton.Priority = {
  PRIMARY: Priority.PRIMARY,
  SECONDARY: Priority.SECONDARY,
};
CircularButton.Type = ControlType;

CircularButton.propTypes = {
  className: Types.string,
  children: Types.string.isRequired,
  disabled: Types.bool,
  icon: Types.element.isRequired,
  onClick: Types.func.isRequired,
  priority: Types.oneOf([CircularButton.Priority.PRIMARY, CircularButton.Priority.SECONDARY]),
  type: Types.oneOf([
    CircularButton.Type.ACCENT,
    CircularButton.Type.POSITIVE,
    CircularButton.Type.NEGATIVE,
  ]),
};

CircularButton.defaultProps = {
  className: '',
  disabled: false,
  priority: CircularButton.Priority.PRIMARY,
  type: CircularButton.Type.ACCENT,
};

export default CircularButton;
