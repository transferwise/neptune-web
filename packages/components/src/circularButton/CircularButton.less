@import './_button-label-states';

.np-circular-btn {
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  cursor: pointer;
  position: relative;

  input[type='button'] {
    min-height: var(--size-40);
    width: var(--size-40);
    height: var(--size-40);
    padding: 0;
    border-radius: 50%;
    margin-bottom: var(--size-8);
  }

  .tw-icon {
    position: absolute;
    top: 8px;
    pointer-events: none;
    color: white;
    transition: color 0.15s ease-in-out;
    width: 100%;
    left: 0;
    > svg {
      margin: 0 auto;
    }
  }
}

.np-circular-btn__label {
  font-weight: 600;
  font-family: inherit;
  transition: color 0.15s ease-in-out;
}

._button-label-states(np-circular-btn, accent,
var(--color-control-accent),
var(--color-control-accent-hover),
var(--color-control-accent-active));

._button-label-states(np-circular-btn, positive,
var(--color-control-positive),
var(--color-control-positive-hover),
var(--color-control-positive-active));

._button-label-states(np-circular-btn, negative,
var(--color-control-negative),
var(--color-control-negative-hover),
var(--color-control-negative-active));

// Boost specificity to ensure rules apply to .secondary buttons
.np-circular-btn.disabled.disabled {
  cursor: not-allowed;

  input[type='button']:disabled + .tw-icon,
  input[type='button']:disabled ~ .np-circular-btn__label {
    color: var(--color-text-inactive);
  }

  &.secondary input[type='button'] {
    border-color: var(--color-text-inactive);
  }
}
