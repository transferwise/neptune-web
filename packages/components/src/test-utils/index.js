import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
