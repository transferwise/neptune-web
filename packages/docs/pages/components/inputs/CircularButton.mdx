import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import { Freeze, Plus, Bin } from '@transferwise/icons';
import { CircularButton } from '@transferwise/components';
import code from '../../../liveEditorCode/circularButton.code';

<LiveEditorBlock code={code} scope={{ CircularButton, Freeze }} />
<GeneratePropsTable componentName="CircularButton" />

## Variants

<div className="scroll-table">
  <table className="docs-table table">
    <thead>
      <tr>
        <th>Type</th>
        <th>Priority</th>
        <th>Preview</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Accent</td>
        <td>Primary</td>
        <td>
          <CircularButton
            icon={<Freeze />}
            type={CircularButton.Type.ACCENT}
            onClick={() => console.log('Button clicked')}
          >
            Freeze
          </CircularButton>
        </td>
      </tr>
      <tr>
        <td>Accent</td>
        <td>Secondary</td>
        <td>
          <CircularButton
            icon={<Freeze />}
            type={CircularButton.Type.ACCENT}
            priority={CircularButton.Priority.SECONDARY}
            onClick={() => console.log('Button clicked')}
          >
            Freeze
          </CircularButton>
        </td>
      </tr>
      <tr>
        <td>Positive</td>
        <td>Primary</td>
        <td>
          <CircularButton
            icon={<Plus />}
            type={CircularButton.Type.POSITIVE}
            onClick={() => console.log('Button clicked')}
          >
            Add money
          </CircularButton>
        </td>
      </tr>
      <tr>
        <td>Positive</td>
        <td>Secondary</td>
        <td>
          <CircularButton
            icon={<Plus />}
            type={CircularButton.Type.POSITIVE}
            priority={CircularButton.Priority.SECONDARY}
            onClick={() => console.log('Button clicked')}
          >
            Add money
          </CircularButton>
        </td>
      </tr>
      <tr>
        <td>Negative</td>
        <td>Primary</td>
        <td>
          <CircularButton
            icon={<Bin />}
            type={CircularButton.Type.NEGATIVE}
            onClick={() => console.log('Button clicked')}
          >
            Delete
          </CircularButton>
        </td>
      </tr>
      <tr>
        <td>Negative</td>
        <td>Secondary</td>
        <td>
          <CircularButton
            icon={<Bin />}
            type={CircularButton.Type.NEGATIVE}
            priority={CircularButton.Priority.SECONDARY}
            onClick={() => console.log('Button clicked')}
          >
            Delete
          </CircularButton>
        </td>
      </tr>
    </tbody>
  </table>
</div>

export const meta = {
  name: 'CircularButton',
  badge: {
    type: 'new',
    expiryDate: '14-04-2021',
  },
};
