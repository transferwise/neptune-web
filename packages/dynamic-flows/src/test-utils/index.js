import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
