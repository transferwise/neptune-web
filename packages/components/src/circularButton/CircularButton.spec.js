import React from 'react';
import '@testing-library/jest-dom';
import { Plus } from '@transferwise/icons';
import { render, screen, userEvent } from '../test-utils';

import CircularButton from './CircularButton';

const { ACCENT, POSITIVE, NEGATIVE } = CircularButton.Type;
const { PRIMARY, SECONDARY } = CircularButton.Priority;

describe('CircularButton', () => {
  const props = {
    children: 'Add money',
    icon: <Plus />,
    onClick: jest.fn(),
  };

  describe('defaults', () => {
    beforeEach(() => {
      render(<CircularButton {...props} />);
    });

    it('renders the text', () => {
      screen.getByText('Add money');
    });

    it('renders the provided icon', () => {
      const icon = document.querySelector('.tw-icon');
      expect(icon).toBeInTheDocument();
    });

    it('is not disabled', () => {
      expect(screen.getByRole('button')).not.toBeDisabled();
    });

    it('renders a button of type accent and priority primary', () => {
      expect(render(<CircularButton {...props} />).container).toMatchSnapshot();
    });
  });

  describe('button attributes', () => {
    it('disables the button if disabled', () => {
      render(<CircularButton {...props} disabled />);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('passes through custom classes if set', () => {
      render(<CircularButton {...props} className="have-a-nice-day-:)" />);
      const label = document.querySelector('label');

      expect(label).toHaveClass('have-a-nice-day-:)');
    });
  });

  describe('onClick', () => {
    it('calls onClick when clicked', () => {
      const onClick = jest.fn();
      render(<CircularButton {...props} onClick={onClick} />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('does not call onClick when clicked if disabled', () => {
      const onClick = jest.fn();
      render(<CircularButton {...props} onClick={onClick} disabled />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(0);
    });
  });

  describe('types', () => {
    it('renders accent buttons', () => {
      expect(render(<CircularButton {...props} type={ACCENT} />).container).toMatchSnapshot();
    });

    it('renders positive buttons', () => {
      expect(render(<CircularButton {...props} type={POSITIVE} />).container).toMatchSnapshot();
    });

    it('renders negative buttons', () => {
      expect(render(<CircularButton {...props} type={NEGATIVE} />).container).toMatchSnapshot();
    });
  });

  describe('priorities', () => {
    it('renders primary buttons', () => {
      [ACCENT, POSITIVE, NEGATIVE].forEach((type) =>
        expect(
          render(<CircularButton {...props} priority={PRIMARY} type={type} />).container,
        ).toMatchSnapshot(),
      );
    });

    it('renders secondary buttons', () => {
      [ACCENT, POSITIVE, NEGATIVE].forEach((type) =>
        expect(
          render(<CircularButton {...props} priority={SECONDARY} type={type} />).container,
        ).toMatchSnapshot(),
      );
    });
  });
});
