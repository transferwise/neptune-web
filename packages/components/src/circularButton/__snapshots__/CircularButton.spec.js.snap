// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CircularButton defaults renders a button of type accent and priority primary 1`] = `
<div>
  <label
    class="np-circular-btn primary accent  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-accent btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders primary buttons 1`] = `
<div>
  <label
    class="np-circular-btn primary accent  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-accent btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders primary buttons 2`] = `
<div>
  <label
    class="np-circular-btn primary positive  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-positive btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders primary buttons 3`] = `
<div>
  <label
    class="np-circular-btn primary negative  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-negative btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders secondary buttons 1`] = `
<div>
  <label
    class="np-circular-btn secondary accent  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-accent btn-priority-2"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders secondary buttons 2`] = `
<div>
  <label
    class="np-circular-btn secondary positive  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-positive btn-priority-2"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton priorities renders secondary buttons 3`] = `
<div>
  <label
    class="np-circular-btn secondary negative  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-negative btn-priority-2"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton types renders accent buttons 1`] = `
<div>
  <label
    class="np-circular-btn primary accent  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-accent btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton types renders negative buttons 1`] = `
<div>
  <label
    class="np-circular-btn primary negative  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-negative btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;

exports[`CircularButton types renders positive buttons 1`] = `
<div>
  <label
    class="np-circular-btn primary positive  "
  >
    <input
      aria-label="Add money"
      class="btn np-btn btn-positive btn-priority-1"
      type="button"
    />
    <span
      aria-hidden="true"
      class="tw-icon tw-icon-plus "
      role="presentation"
    >
      <svg
        fill="currentColor"
        focusable="false"
        height="24"
        width="24"
      >
        <path
          clip-rule="evenodd"
          d="M11 22V2h2v20h-2z"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M22 13H2v-2h20v2z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
    <span
      class="np-circular-btn__label"
    >
      Add money
    </span>
  </label>
</div>
`;
